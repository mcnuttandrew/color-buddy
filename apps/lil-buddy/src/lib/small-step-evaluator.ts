import { LLTypes, linter, Environment } from "color-buddy-palette-lint";
import { Color } from "color-buddy-palette";
import type { Palette } from "color-buddy-palette";
type LLNode = InstanceType<(typeof LLTypes)["LLNode"]>;

export function evaluateNode(
  node: any,
  inducedVariables: Record<string, Color>,
  pal: Palette
) {
  const opts = { debugParse: false, debugEval: false, debugCompare: false };
  let newEnv = new Environment(pal, {}, opts, {});

  newEnv = Object.entries(inducedVariables).reduce((acc, [key, value]) => {
    const newVal = new LLTypes.LLValue(
      new LLTypes.LLColor(value, value.toHex())
    );
    return acc.set(key, newVal);
  }, newEnv);
  return node.evaluate(newEnv);
}

function isValue(node: any) {
  switch (node.nodeType) {
    case "value":
      return true;
    case "expression":
    case "node":
      return isValue(node.value);
    case "bool":
    case "color":
    case "number":
    case "variable":
    case undefined:
      return true;
    default:
      return false;
  }
}
function subTreeIsPureOp(node: any): boolean {
  switch (node.nodeType) {
    case "pairFunction":
    case "predicate":
      return isValue(node.left) && isValue(node.right);
    case "conjunction":
      if (node.type === "not") {
        return isValue(node.children[0]);
      }
      return node.children.every((x: any) => subTreeIsPureOp(x));
    case "array":
      return node.children.every((x: any) => subTreeIsPureOp(x));
    case "node":
    case "expression":
      return subTreeIsPureOp(node.value);
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
      return false;
    case "valueFunction":
      return isValue(node.input);
    // todo
    case "quantifier":
    case "aggregate":
    case "map":
    default:
      return false;
  }
}
function traverseAndMaybeExecute(
  node: any,
  inducedVariables: Record<string, Color>,
  pal: Palette
): {
  result: any;
  didEval: boolean;
} {
  const thisIsPureOp = subTreeIsPureOp(node);
  if (thisIsPureOp) {
    const result = evaluateNode(node, inducedVariables, pal).result;
    const astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
    return { result: astResult, didEval: true };
  }
  //   let updatedNode = node.deepCopy();
  let updatedNode = Object.assign({}, node);
  switch (node.nodeType) {
    case "pairFunction":
    case "predicate":
      const leftTraverse = traverseAndMaybeExecute(
        node.left,
        inducedVariables,
        pal
      );
      if (leftTraverse.didEval) {
        updatedNode.left = leftTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      const rightTraverse = traverseAndMaybeExecute(
        node.right,
        inducedVariables,
        pal
      );
      if (rightTraverse) {
        updatedNode.right = rightTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      return { result: updatedNode, didEval: false };
    case "conjunction":

    case "array":
      const newChildren = [];
      let found = false;
      for (let idx = 0; idx < node.children.length; idx++) {
        if (found) {
          newChildren.push(node.children[idx]);
          continue;
        }
        const child = node.children[idx];
        const childResult = traverseAndMaybeExecute(
          child,
          inducedVariables,
          pal
        );
        newChildren.push(childResult.result);
        found = childResult.didEval;
      }
      node.children = newChildren;
      return { result: node, didEval: found };
    case "node":
    case "expression":
      return traverseAndMaybeExecute(node.value, inducedVariables, pal);
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
    case undefined:
      return { result: node, didEval: false };
    case "valueFunction":
      const arg = traverseAndMaybeExecute(node.input, inducedVariables, pal);
      if (arg.didEval) {
        updatedNode.input = arg.result;
        return { result: updatedNode, didEval: true };
      } else {
        return { result: updatedNode, didEval: false };
      }

    case "quantifier":
    case "aggregate":
    case "map":
    default:
      console.log(node.nodeType, " not implemented yet", node);
      throw new Error(`${node.nodeType} not implemented yet`, node);
      return { result: node, didEval: false };
  }
}
export function generateEvaluations(
  node: LLNode,
  inducedVariables: Record<string, Color>,
  pal: Palette
): LLNode[] {
  const evalLog = [node];
  //   let currentNode = node.deepCopy();
  let currentNode = Object.assign({}, node);
  while (!subTreeIsPureOp(currentNode)) {
    const result = traverseAndMaybeExecute(currentNode, inducedVariables, pal);
    evalLog.push(result.result);
    currentNode = result.result;
  }
  const result = evaluateNode(node, inducedVariables, pal).result;
  const astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
  evalLog.push(astResult);
  return evalLog;
}
