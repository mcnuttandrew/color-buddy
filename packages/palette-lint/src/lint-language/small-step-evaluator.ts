import { Color, makePalFromString } from "color-buddy-palette";
import type { Palette } from "color-buddy-palette";
import { LLTypes, Environment } from "./lint-language";

type LLNode = InstanceType<(typeof LLTypes)["LLNode"]>;
export type InducedVariables = Record<string, Color | number>;

export function rewriteQuantifiers(node: any) {
  if (!node || !node.nodeType) {
    return node;
  }
  const newNode = copy(node);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const left = rewriteQuantifiers(newNode.left);
      const right = rewriteQuantifiers(newNode.right);
      newNode.left = left;
      newNode.right = right;
      break;
    case "aggregate":
    case "array":
    case "conjunction":
    case "map":
      const children = node.children;
      if (Array.isArray(children)) {
        newNode.children = children.map(rewriteQuantifiers);
      } else {
        newNode.children = rewriteQuantifiers(children);
      }
      break;
    case "node":
    case "expression":
      newNode.value = rewriteQuantifiers(node.value);
      break;
    case "boolFunction":
    case "valueFunction":
      newNode.input = rewriteQuantifiers(node.input);
      break;
    case "quantifier":
      if (node.varbs.length === 1) {
        newNode.where = rewriteQuantifiers(node.where);
        newNode.predicate = rewriteQuantifiers(node.predicate);
      } else {
        const [head, ...tail] = node.varbs;
        const newSubNode = copy(node);
        newSubNode.where = rewriteQuantifiers(node.where);
        newSubNode.predicate = node.predicate;
        newSubNode.varbs = tail;
        newNode.where = undefined;
        newNode.varbs = [head];
        newNode.predicate = rewriteQuantifiers(newSubNode);
      }
      break;
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
    default:
      break;
  }
  return newNode;
}

function clearQuantifierResults(node: any) {
  if (!node) {
    return node;
  }
  if (!node.nodeType && !node.quant) {
    return node;
  } else if (node.quant) {
    return new LLTypes.LLBool(node.quantifierResult);
  }
  const newNode = copy(node);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const left = clearQuantifierResults(newNode.left);
      const right = clearQuantifierResults(newNode.right);
      newNode.left = left;
      newNode.right = right;
      break;
    case "aggregate":
    case "array":
    case "conjunction":
    case "map":
      const children = node.children;
      if (Array.isArray(children)) {
        newNode.children = children.map(clearQuantifierResults);
      } else {
        newNode.children = clearQuantifierResults(children);
      }
      break;
    case "node":
    case "expression":
      newNode.value = clearQuantifierResults(node.value);
      break;
    case "valueFunction":
    case "boolFunction":
      newNode.input = clearQuantifierResults(node.input);
      break;
    case "quantifier":
      newNode.where = clearQuantifierResults(node.where);
      newNode.predicate = clearQuantifierResults(node.predicate);
      break;
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
    default:
      break;
  }
  return newNode;
}

function checkWhere(
  node: any,
  color: Color,
  varb: string,
  pal: Palette,
  inducedVariables: InducedVariables,
  index: number
): boolean {
  if (!node) return true;
  const result = evaluateNode(
    node,
    { ...inducedVariables, [varb]: color, [`index(${varb})`]: index + 1 },
    pal
  );
  return result.result;
}

function getValues(node: any, pal: Palette) {
  if (node?.input?.value === "colors") {
    return [...pal.colors];
  }
  if (
    Array.isArray(node?.input?.children) &&
    typeof node.input.children?.at(0)?.constructorString
  ) {
    return makePalFromString(
      node.input.children.map((x: any) => x.constructorString)
    ).colors;
  }

  return [];
}

const toVal = (x: Color | number) => {
  if (typeof x === "number") {
    return new LLTypes.LLValue(new LLTypes.LLNumber(x));
  } else if (typeof x === "string") {
    return new LLTypes.LLValue(
      new LLTypes.LLColor(Color.colorFromHex(x, "lab"), x)
    );
  }
  return new LLTypes.LLValue(new LLTypes.LLColor(x, x.toHex()));
};
const evalCache = new Map<string, any>();
function evaluateNode(
  node: any,
  inducedVariables: InducedVariables,
  pal: Palette
) {
  const serializedNode = JSON.stringify(node);
  const canUseCache = !serializedNode.includes("colors");
  const serializeKey = JSON.stringify({
    serializedNode,
    inducedVariables,
    background: pal.background.toHex(),
  });
  if (canUseCache && evalCache.has(serializeKey)) {
    return evalCache.get(serializeKey);
  }

  const opts = { debugParse: false, debugEval: false, debugCompare: false };

  const newEnv = Object.entries(inducedVariables).reduce(
    (acc, [key, value]) => acc.set(key, toVal(value)),
    new Environment(pal, {}, opts, {})
  );
  const cleanedNode = clearQuantifierResults(node);
  const result = cleanedNode.evaluate(newEnv);
  if (canUseCache) {
    evalCache.set(serializeKey, result);
  }
  return result;
}

function isValue(node: any) {
  switch (node.nodeType) {
    case "value":
      return true;
    case "expression":
    case "node":
      return isValue(node.value);
    case "bool":
    case "color":
    case "number":
    case "variable":
    case undefined:
      return true;
    case "array":
      return node.children.every(isValue);
    default:
      return false;
  }
}

function subTreeIsPureOp(
  node: any,
  inducedVariables: InducedVariables
): boolean {
  node.inducedVariables = toHexes(inducedVariables);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      return isValue(node.left) && isValue(node.right);
    case "aggregate":
    case "array":
    case "map":
      if (node.children?.nodeType === "variable") {
        return true;
      }
      const children = node.children;
      if (Array.isArray(children)) {
        return children.every((x: any) => isValue(x));
      } else {
        // for variables and such
        return isValue(children);
      }
    case "node":
    case "expression":
      return subTreeIsPureOp(node.value, inducedVariables);
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
      return false;
    case "boolFunction":
    case "valueFunction":
      return isValue(node.input);
    case "quantifier":
      return false;
    default:
      return false;
  }
}

function mapObject(obj: any, fn: (k: string, v: any) => any) {
  return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, fn(k, v)]));
}
function toHexes(obj: any) {
  return mapObject(obj, (k, v) => {
    if (typeof v === "number") {
      return v;
    }
    if (!v.toHex) {
      return v;
    }
    return v.toHex();
  });
}

function stripOutIds(str: string): string {
  return str.replace(/"id":\d+/g, '"id":0');
}

let counter = 0;
function traverseAndMaybeExecute(
  node: any,
  inducedVariables: InducedVariables,
  pal: Palette
): {
  result: any;
  didEval: boolean;
} {
  counter++;
  if (counter > 500) {
    throw new Error("Too many iterations");
  }
  const thisIsPureOp = subTreeIsPureOp(node, inducedVariables);
  const allChildrenPureOps = Array.isArray(node?.children)
    ? (node?.children || []).every((x: any) => isValue(x))
    : false;
  if (thisIsPureOp || allChildrenPureOps) {
    const preKey = JSON.stringify(node.copy());
    const result = evaluateNode(node, inducedVariables, pal).result;
    let astResult;
    if (Array.isArray(result)) {
      astResult = LLTypes.LLValueArray.tryToConstruct(result, {} as any);
    } else {
      astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
    }
    const postKey = JSON.stringify(astResult.copy());
    return {
      result: astResult,
      didEval: stripOutIds(preKey) !== stripOutIds(postKey),
    };
  }

  let updatedNode = copy(node);
  subTreeIsPureOp(updatedNode, inducedVariables);
  updatedNode.inducedVariables = toHexes({ ...inducedVariables });

  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const leftTraverse = traverseAndMaybeExecute(
        updatedNode.left,
        inducedVariables,
        pal
      );
      if (leftTraverse.didEval) {
        updatedNode.left = leftTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      const rightTraverse = traverseAndMaybeExecute(
        updatedNode.right,
        inducedVariables,
        pal
      );
      if (rightTraverse.didEval) {
        updatedNode.right = rightTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      return { result: updatedNode, didEval: false };
    case "conjunction":
    case "array":
    case "aggregate":
    case "map":
      const children = updatedNode.children;
      if (Array.isArray(children)) {
        const newChildren = [];
        let found = false;
        for (let idx = 0; idx < children.length; idx++) {
          if (found) {
            newChildren.push(children[idx]);
            continue;
          }
          const child = children[idx];
          const childResult = traverseAndMaybeExecute(
            child,
            inducedVariables,
            pal
          );
          newChildren.push(childResult.result);
          found = childResult.didEval;
        }
        updatedNode.children = newChildren;
        return { result: updatedNode, didEval: found };
      } else {
        const childResult = traverseAndMaybeExecute(
          children,
          inducedVariables,
          pal
        );
        updatedNode.children = childResult.result;
        return { result: updatedNode, didEval: childResult.didEval };
      }
    case "node":
    case "expression":
      return traverseAndMaybeExecute(node.value, inducedVariables, pal);
    case "bool":
    case "color":
    case "number":
    case "value":
    case undefined:
      return { result: updatedNode, didEval: false };
    case "variable":
      return { result: updatedNode, didEval: false };
    case "boolFunction":
    case "valueFunction":
      const arg = traverseAndMaybeExecute(
        updatedNode.input,
        inducedVariables,
        pal
      );
      if (arg.didEval) {
        updatedNode.input = arg.result;
        return { result: updatedNode, didEval: true };
      } else {
        return { result: updatedNode, didEval: false };
      }

    case "quantifier":
      const values = getValues(updatedNode, pal);
      // const results = [];
      const results = values.map((color, i) => {
        const whereResult = checkWhere(
          updatedNode.where,
          color,
          updatedNode.varbs[0],
          pal,
          { ...updatedNode.inducedVariables },
          i
        );
        if (!whereResult) {
          return { result: "WHERE SKIP", didEval: true, color: color.toHex() };
        }
        const updatedVariables = {
          ...updatedNode.inducedVariables,
          [updatedNode.varbs[0]]: color,
          [`index(${node.varbs[0]})`]: i + 1,
        };
        const evals = smallStepEvaluator(
          updatedNode.predicate.value || updatedNode.predicate,
          { ...updatedVariables },
          pal
        );
        return {
          color: color.toHex(),
          evals: [...evals],
        };
      });
      const quantifierResult = evaluateNode(node, inducedVariables, pal).result;
      return {
        result: {
          results,
          quant: updatedNode.type,
          varb: updatedNode.varbs[0],
          quantifierResult,
        },
        didEval: true,
      };

    default:
      console.log(node.nodeType, " not implemented yet", node);
      throw new Error(`${node.nodeType} not implemented yet`, node);
    // return { result: node, didEval: false };
  }
}

function copy(node: any) {
  if (node.copy) {
    return node.copy();
  }
  return JSON.parse(JSON.stringify(node));
}

export function smallStepEvaluator(
  node: LLNode,
  inducedVariables: InducedVariables,
  pal: Palette,
  init?: boolean
): LLNode[] {
  const serializedNode = JSON.stringify(node);
  const canUseCache = !serializedNode.includes("colors");
  const serializeKey = JSON.stringify({
    serializedNode,
    inducedVariables,
    background: pal.background.toHex(),
    generate: "generate",
  });
  if (canUseCache && evalCache.has(serializeKey)) {
    console.log("cache hit");
    return evalCache.get(serializeKey);
  }
  if (init) {
    counter = 0;
  }
  let nodeCopy = copy(node);
  // weird hack to insert the induced variables over everything
  subTreeIsPureOp(nodeCopy, inducedVariables);
  const evalLog = node.nodeType === "quantifier" ? [] : [nodeCopy];
  let currentNode = copy(node);
  subTreeIsPureOp(node, inducedVariables);

  while (
    !subTreeIsPureOp(currentNode, { ...inducedVariables }) &&
    !isValue(currentNode)
  ) {
    const result = traverseAndMaybeExecute(
      currentNode,
      { ...inducedVariables },
      pal
    );
    result.result.inducedVariables = toHexes({ ...inducedVariables });

    evalLog.push(result.result);
    currentNode = result.result;
  }
  // get the final result
  const result = evaluateNode(node, { ...inducedVariables }, pal).result;
  const astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
  evalLog.push(astResult);
  // check if the last two items are the same, if so, remove the last one
  if (
    evalLog.length > 1 &&
    prepStr(evalLog[evalLog.length - 1]) ===
      prepStr(evalLog[evalLog.length - 2])
  ) {
    evalLog.pop();
  }
  if (canUseCache) {
    evalCache.set(serializeKey, evalLog);
  }
  return evalLog;
}

const prepStr = (node: any) => stripOutIds(JSON.stringify(copy(node)));
