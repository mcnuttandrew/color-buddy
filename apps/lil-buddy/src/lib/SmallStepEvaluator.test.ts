import { expect, test } from "vitest";
import { makePalFromString } from "color-buddy-palette";
import { GenerateAST } from "color-buddy-palette-lint";
import { generateEvaluations } from "./small-step-evaluator";

test("SmallStepEvaluator works", () => {
  const exampleNode = {
    ">": {
      left: { deltaE: { left: "a", right: "b" }, algorithm: "2000" },
      right: 10,
    },
  };
  const ast = (GenerateAST(exampleNode as any).value as any).children[0] as any;
  const pal = makePalFromString(["red", "green"]);
  const result = generateEvaluations(
    ast,
    { a: pal.colors[0], b: pal.colors[1] },
    pal
  );
  expect(result).toMatchSnapshot();
});

test("SmallStepEvaluator works with smaller example", () => {
  const smallExampleNode = { ">": { left: 11, right: 10 } };
  const ast = (GenerateAST(smallExampleNode as any).value as any)
    .children[0] as any;
  const result = generateEvaluations(
    ast,
    {},
    makePalFromString(["red", "green"])
  );
  expect(result).toMatchSnapshot();
});

test.only("SmallStepEvaluator works with small not example", () => {
  const smallExampleNode = { not: { ">": { left: 11, right: 10 } } };
  const ast = (GenerateAST(smallExampleNode as any).value as any)
    .children[0] as any;
  const result = generateEvaluations(
    ast,
    {},
    makePalFromString(["red", "green"])
  );
  expect(result).toMatchSnapshot();
});
