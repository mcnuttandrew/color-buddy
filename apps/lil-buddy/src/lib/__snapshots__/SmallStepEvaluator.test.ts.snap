// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SmallStepEvaluator works 1`] = `
[
  LLExpression {
    "nodeType": "expression",
    "value": LLPredicate {
      "left": LLPairFunction {
        "left": LLVariable {
          "nodeType": "variable",
          "value": "a",
        },
        "nodeType": "pairFunction",
        "params": {
          "algorithm": "2000",
        },
        "right": LLVariable {
          "nodeType": "variable",
          "value": "b",
        },
        "type": "deltaE",
      },
      "nodeType": "predicate",
      "right": LLNumber {
        "nodeType": "number",
        "value": 10,
      },
      "threshold": undefined,
      "type": ">",
    },
  },
  {
    "left": LLNumber {
      "nodeType": "number",
      "value": 70.2365636120491,
    },
    "nodeType": "predicate",
    "right": LLNumber {
      "nodeType": "number",
      "value": 10,
    },
    "threshold": undefined,
    "type": ">",
  },
  LLBool {
    "nodeType": "bool",
    "value": true,
  },
]
`;

exports[`SmallStepEvaluator works with small not example 1`] = `
[
  LLExpression {
    "nodeType": "expression",
    "value": LLConjunction {
      "children": [
        LLBool {
          "nodeType": "bool",
          "value": true,
        },
      ],
      "nodeType": "conjunction",
      "type": "not",
    },
  },
  LLConjunction {
    "children": [
      LLBool {
        "nodeType": "bool",
        "value": true,
      },
    ],
    "nodeType": "conjunction",
    "type": "not",
  },
  LLBool {
    "nodeType": "bool",
    "value": false,
  },
]
`;

exports[`SmallStepEvaluator works with smaller example 1`] = `
[
  LLExpression {
    "nodeType": "expression",
    "value": LLPredicate {
      "left": LLNumber {
        "nodeType": "number",
        "value": 11,
      },
      "nodeType": "predicate",
      "right": LLNumber {
        "nodeType": "number",
        "value": 10,
      },
      "threshold": undefined,
      "type": ">",
    },
  },
  LLBool {
    "nodeType": "bool",
    "value": true,
  },
]
`;
